### PR TITLE
Experimental Typescript support

### PR DESCRIPTION
**Pull request by stanley-cheung**
For #255 

I am going to try to expand on this. For my very initial understanding, this seems to work.

Feedback highly appreciated!

cc: @jonahbron

### CODE REVIEW

**Comment by stanley-cheung** [comment]
So far, this PR demos how a typescript project can make use of the `index.d.ts` typings for the `grpc-web` module. I hand-wrote an `echo_grpc_web_pb.ts` file that could be generated from a future grpc-web x TS protoc plugin. To test whether the `index.d.ts` typings do matter, I flip, let's say, a `hostname` parameter from a `string` to a `number`. Now the `tsc` command will fail because a string is being passed.

**Comment by jonahbron** [comment]
That looks right to me.  One alternative to the `echo_grpc_web_pb.ts` file I've seen in [ts-protoc-gen](https://github.com/improbable-eng/ts-protoc-gen) is producing .d.ts for the generated files too.  That way the JS generator can still be used as-is.

File: packages/grpc-web/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  export class ClientReadableStream {
+    on (type: string,
+        callback: (...args: any[]) => any): ClientReadableStream;
+    cancel (): void;
+  }
+
+  export interface GrpcWebClientBaseOptions {
+    format?: string;
```
Comment (aberasarte) [inline packages/grpc-web/index.d.ts]: I think that we could add the `suppressCorsPreflight` option as well: `suppressCorsPreflight?: boolean`

Comment (stanley-cheung) [reply packages/grpc-web/index.d.ts]: Will do.

**Comment by stanley-cheung** [comment]
> That looks right to me. One alternative to the echo_grpc_web_pb.ts file I've seen in ts-protoc-gen is producing .d.ts for the generated files too. That way the JS generator can still be used as-is.

Will try that. Thanks.

**Comment by stanley-cheung** [comment]
Added a full Typescript example up to this point.

`ts-example/EchoServiceClient.ts`: a hand-written would-be full-TS codegen output

`ts-example/echo_grpc_web_pb.d.ts`: .d.ts typings that is meant to work alongside existing `echo_grpc_web_pb.js`  CommonJS codegen output
`ts-example/echo_pb.d.ts`: .d.ts typings for protobuf message types, meant to work alongside `echo_pb.js` like above

`ts-example/client.ts`: TS client app code

Uncomment either Line 25 or Line 26 to try between the two modes above:
```
import {EchoServiceClient} from './echo_grpc_web_pb';   // .d.ts typings for existing JS codegen output
// import {EchoServiceClient} from './EchoServiceClient';  // full-TS codegen output
```

**Comment by stanley-cheung** [comment]
Added 2 modes to the `protoc-gen-grpc-web` protoc plugin

 - `import_style=commonjs+dts`: this will generate `.d.ts` typings for both the `<proto>_grpc_web_pb.js` service stub and the `<proto>_pb.js` proto message classes
 - `import_style=typescript`: this will generate `<proto>ServiceClientPb.ts` typescript service stub

**Comment by stanley-cheung** [comment]
Squashing commits. Got internal reviewed. Will merge this soon.

**Pull request merged**